A basic progress meter for walker-based tools

droazen opened the pull request:
-Prints the current locus, the elapsed time, number of records processed,
 and the rate at which records are being processed.

-Hooked up for ReadWalkers, VariantWalkers, and IntervalWalkers

-A new command-line arg in GATKTool allows control over the frequency of
 progress meter updates

-Tweaked the log4j output format to create more screen space for logger output.

Resolves #974 (for alpha purposes)

droazen commented:
For @akiezun to review

akiezun commented on src/main/java/org/broadinstitute/hellbender/engine/ProgressMeter.java:
```diff
@@ -0,0 +1,200 @@
+ *
+ * All output is made at INFO level via log4j.
+ */
+public class ProgressMeter {
```
make class final or document which methods are to be extended or overriden and in what way. I suggest just making final to make the intentions clear.

droazen replied:
Made final

akiezun commented on src/main/java/org/broadinstitute/hellbender/engine/ProgressMeter.java:
```diff
@@ -0,0 +1,200 @@
+     * The genomic location of the most recently processed record, or null if the most recent record had no location.
+     * Updated only when we actually output a line to the logger.
+     */
+    private SimpleInterval currentLocus = null;
```
can this be a `Locatable`?

droazen replied:
Sure

akiezun commented on src/main/java/org/broadinstitute/hellbender/engine/ProgressMeter.java:
```diff
@@ -0,0 +1,200 @@
+     */
+    @VisibleForTesting
+    double elapsedTimeInMinutes() {
+        return (currentTime - startTime) / 1000.0 / 60.0;
```
this calculation shows that the time is measures in milliseconds. Please a comment and/or change names of `currentTime` and `startTime` so say it is so. Also, I'd just extract 1000.0 * 60.0 as MILLISECONDS_PER_MINUTE to clarify

droazen replied:
Done and done

akiezun commented on src/test/java/org/broadinstitute/hellbender/engine/ProgressMeterUnitTest.java:
```diff
@@ -0,0 +1,124 @@
+    public Object[][] getElapsedTimeInMinutesTestData() {
+        return new Object[][] {
+                // 600,000ms = 10 minutes
+                { new ListBasedTimeFunction(Arrays.asList(1000l, 601000l)), 10.0 },
```
I'd maybe prefer `10 * 60000l + 1000l` for clarity (and same elsewhere, ie factor out the number of minutes and add 1000 separately)

droazen replied:
Done

akiezun commented:
very small comments to consider. LGTM otherwise

akiezun commented on src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java:
```diff
@@ -268,6 +279,8 @@ protected void onStartup() {
 
         checkToolRequirements();
+
+        progressMeter = new ProgressMeter(recordsPerProgressUpdate);
```
I think this should be extracted out to a method that subclasses can override. Maybe `initializeProgressMeter` or something. That way, subclasses can have a say in what the default rate of updates is (and not delegate all to commandline options). For example perhaps the default rate for variant walkers should be smaller than read walkers etc.

droazen replied:
After seeing how different the speed of the progress meter is between read and variant walkers, I am actually going to switch to a time-based model.

droazen commented:
@akiezun I've addressed your comments, and also switched to a time-based model where we keep track of the time elapsed since last progress output (polling the time every 1000 records to cut down on system call overhead), and output status every X seconds. Please re-review.

akiezun commented:
LGTM
